### PR TITLE
New Actor - aws.iam.Role

### PR DESCRIPTION
**Pull request by diranged**
This initial work just allows us to create a role and manage its inline
policies. I still have work to do to make sure that the role is usable
by other instance profiles.

Reviewer: @siminm

### CODE REVIEW

File: kingpin/actors/aws/iam/entities.py [hunk 1]
```diff
@@ -599,3 +611,148 @@ def __init__(self, *args, **kwargs):
...
+
+class Role(EntityBaseActor):
+
+    """Manages an IAM Role.
+
+    This actor manages the state of an Amazon IAM Role. It ensures that the
+    role either exists or does not. It also updates any settings for the role
+    that are different from the passed in options.
+
+    Currently we can:
```
Comment (siminm) [inline kingpin/actors/aws/iam/entities.py]: What's the point of this block vs the description right above this?

File: kingpin/actors/aws/iam/entities.py [hunk 2]
```diff
@@ -599,3 +611,148 @@ def __init__(self, *args, **kwargs):
...
+
+class InstanceProfile(EntityBaseActor):
+
+    """Manages an IAM Instance Profile.
+
+    This actor manages the state of an Amazon IAM Instance Profile. It ensures
+    that the profile either exists or does not. It also updates any settings
+    for the profile that are different from the passed in options.
+
+    Currently we can:
```
Comment (siminm) [inline kingpin/actors/aws/iam/entities.py]: Ditto

File: kingpin/actors/aws/iam/test/integration_entities.py [hunk 3]
```diff
@@ -114,6 +114,112 @@ def integration_02b_create_group(self):
...
+             'inline_policies': self.inline_policies,
+             'inline_policies_purge': True},
+            dry=False)
+
+        yield actor.execute()
+
+    # Final cleanup -- delete our test profile!
+    @attr('aws', 'integration')
+    @testing.gen_test(timeout=60)
+    def integration_09_ensure_profile_absent(self):
```
Comment (siminm) [inline kingpin/actors/aws/iam/test/integration_entities.py]: 09?

File: kingpin/actors/aws/iam/test/integration_entities.py [hunk 4]
```diff
@@ -114,6 +114,112 @@ def integration_02b_create_group(self):
...
+
+    integration = True
+
+    name = 'kingpin-integration-test'
+    inline_policies = [
+        'examples/aws.iam.user/s3_example.json'
+    ]
+    region = 'us-east-1'
+
+    # Not really a test - this is just a state cleaner. Ensure that we start
```
Comment (siminm) [inline kingpin/actors/aws/iam/test/integration_entities.py]: Technically a valid test - making sure that having an absent profile when it doesn't exist works correctly

**Comment by diranged** [comment]
@siminm i have addressed your comments

File: kingpin/actors/aws/iam/entities.py [hunk 5]
```diff
@@ -525,23 +580,96 @@ def __init__(self, *args, **kwargs):
...
+            groups: The list (or single) of groups to join be members of
+            purge: Whether or not to purge unmanaged targets.
+        """
+        if isinstance(groups, basestring):
+            groups = [groups]
+
+        current_groups = []
+        try:
+            raw = yield self.thread(self.iam_conn.get_groups_for_user, name)
+            resp_key = 'list_groups_for_user_response'
```
Comment (siminm) [inline kingpin/actors/aws/iam/entities.py]: All lines that do not raise the exception should not be inside of the try. Either put them into the `else` part of try/catch or just put them outside of the entire block

File: kingpin/actors/aws/iam/entities.py [hunk 6]
```diff
@@ -525,23 +580,96 @@ def __init__(self, *args, **kwargs):
...
+            result_key = 'list_groups_for_user_result'
+            existing_groups = raw[resp_key][result_key]['groups']
+
+            # Quickly just pull out the friendly names
+            current_groups = [group['group_name'] for group in existing_groups]
```
Comment (siminm) [inline kingpin/actors/aws/iam/entities.py]: If you change this list comprehension to a set-comprehension you can get very easy math out of it. Here's a full blown example!

```python
import boto.iam
ic = boto.iam.connection.IAMConnection()
all_groups = ic.get_groups_for_user('mwise-test-user-1');
all_groups
# 
{u'list_groups_for_user_response': {u'list_groups_for_user_result': {u'groups': [{u'arn': u'arn:aws:iam::364942603424:group/mwise-test-group',
     u'create_date': u'2016-04-22T16:32:04Z',
     u'group_id': u'AGPAIECWMG5SNOZG2GPJY',
     u'group_name': u'mwise-test-group',
     u'path': u'/'},
    {u'arn': u'arn:aws:iam::364942603424:group/mwise-test-group-2',
     u'create_date': u'2016-04-22T16:32:05Z',
     u'group_id': u'AGPAJXSEU7BYYYUNE4E5W',
     u'group_name': u'mwise-test-group-2',
     u'path': u'/'}],
   u'is_truncated': u'false'},
  u'response_metadata': {u'request_id': u'a92cd353-08c7-11e6-8844-2b912bbfcb1f'}}}

current_groups = {g['group_name'] for g in all_groups['list_groups_for_user_response']
                                                     ['list_groups_for_user_result']
                                                     ['groups']}
current_groups
# {u’mwise-test-group', u'mwise-test-group-2'}

ensure_groups = {u'mwise-test-group-2', u'mwise-test-group-3'}

to_create = ensure_groups - current_groups
to_create
#  {u’mwise-test-group-3'}

to_delete = current_groups - ensure_groups
to_delete
# {u’mwise-test-group'}
```

File: kingpin/actors/aws/iam/entities.py [hunk 7]
```diff
@@ -599,3 +728,314 @@ def __init__(self, *args, **kwargs):
...
+        returns:
+            a list of user name strings
+        """
+        users = []
+        try:
+            raw = yield self.thread(self.iam_conn.get_group, name)
+            resp_key = 'get_group_response'
+            result_key = 'get_group_result'
+            users = raw[resp_key][result_key]['users']
+            users = [user['user_name'] for user in users]
```
Comment (siminm) [inline kingpin/actors/aws/iam/entities.py]: Try employing the same pattern as for Group handling

File: kingpin/actors/aws/iam/entities.py [hunk 8]
```diff
@@ -599,3 +728,314 @@ def __init__(self, *args, **kwargs):
...
+        'purge' option was set.
+
+        args:
+          name: the group name
+          force: boolean whether or not to actually force the removal
+        """
+        users = yield self._get_group_users(name)
+
+        if not force and users:
+            self.log.warning(('Will not be able to delete this group ' +
```
Comment (siminm) [inline kingpin/actors/aws/iam/entities.py]: No need for a + here at all. Python has implicit string concatenation.
```python
s = ('hello '
     'world')
```

File: kingpin/actors/aws/iam/entities.py [hunk 9]
```diff
@@ -599,3 +722,314 @@ def __init__(self, *args, **kwargs):
...
+                    'An unexpected API error occurred: %s' % e)
+
+    @gen.coroutine
+    def _ensure_role(self, name, role):
+        """Ensures that an Instance Profile role is set correctly.
+
+        Adds, Deletes or Changes the Role assigned to an Instance Profile.
+
+        args:
+            name: The IAM Instance Profile we're managing
```
Comment (siminm) [inline kingpin/actors/aws/iam/entities.py]: Leave a comment here that this is a string and not an IAM object

**Comment by siminm** [comment]
Let's do this! 👍

**Pull request merged**